File: myExpenses/src/main/java/org/totschnig/myexpenses/service/GenericAlarmReceiver.java
/*******************************************************************************
 * Copyright (c) 2010 Denis Solonenko.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 * <p>
 * Contributors:
 * Denis Solonenko - initial API and implementation
 ******************************************************************************/
package org.totschnig.myexpenses.service;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import org.totschnig.myexpenses.model.Account;

import timber.log.Timber;

public class GenericAlarmReceiver extends BroadcastReceiver {

  static final String BOOT_COMPLETED = "android.intent.action.BOOT_COMPLETED";
  static final String ACCOUNT_CHANGED = "android.accounts.LOGIN_ACCOUNTS_CHANGED";

  @Override
  public void onReceive(Context context, Intent intent) {
    String action = intent.getAction();
    if (BOOT_COMPLETED.equals(action)) {
      requestScheduleAutoBackup(context);
      requestSchedulePlanExecutor(context);
    } else if (ACCOUNT_CHANGED.equals(action)) {
      try {
        Account.checkSyncAccounts(context);
      } catch (Exception e) {
        Timber.e(e);
      }
    }
  }

  private void requestSchedulePlanExecutor(Context context) {
    Intent serviceIntent = new Intent(context, PlanExecutor.class);
    serviceIntent.setAction(PlanExecutor.ACTION_SCHEDULE_EXECUTE_PLANS);
    PlanExecutor.enqueueWork(context, serviceIntent);
  }

  private void requestScheduleAutoBackup(Context context) {
    Intent serviceIntent = new Intent(context, AutoBackupService.class);
    serviceIntent.setAction(AutoBackupService.ACTION_SCHEDULE_AUTO_BACKUP);
    AutoBackupService.enqueueWork(context, serviceIntent);
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/service/GenericAlarmReceiver.kt
package org.totschnig.myexpenses.service

import android.accounts.AccountManager
import android.app.NotificationManager
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.WorkManager
import org.totschnig.myexpenses.MyApplication
import org.totschnig.myexpenses.model.Account
import org.totschnig.myexpenses.preference.PrefKey
import org.totschnig.myexpenses.util.NotificationBuilderWrapper.NOTIFICATION_AUTO_BACKUP
import timber.log.Timber

class GenericAlarmReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        when (intent.action) {
            AccountManager.LOGIN_ACCOUNTS_CHANGED_ACTION -> try {
                if ((context.applicationContext as MyApplication).appComponent.prefHandler().getInt(PrefKey.CURRENT_VERSION, 0) > 0) {
                    Account.checkSyncAccounts(context)
                }
            } catch (e: Exception) {
                Timber.e(e)
            }
            AutoBackupWorker.ACTION_BACKUP_PURGE_CANCEL -> {
                (context.getSystemService(Context.NOTIFICATION_SERVICE) as? NotificationManager)?.cancel(
                    NOTIFICATION_AUTO_BACKUP
                )
            }
            AutoBackupWorker.ACTION_BACKUP_PURGE -> {
                WorkManager.getInstance(context).enqueue(OneTimeWorkRequestBuilder<BackupPurgeWorker>().build())
            }
        }
    }
}